### DISCUSSION TITLE
Using ChatGPT to Develop High Quality Documentation for Freeplane Features

### DISCUSSION THREAD

**Discussion by dpolivaev** [freeplane/freeplane#893] [Announcements]
ChatGPT, currently available as a free preview at https://chat.openai.com/chat, is a unique tool that can help us generate high quality documentation for Freeplane features. While it is not capable of developing its own ideas, it knows how to phrase everything in a way that is easy to understand by most people.

That's why I've been using ChatGPT for all of my new posts. And I think it can be a great resource for documenting Freeplane features as well. We can start by brainstorming a rough structure and ideas, and then use ChatGPT to restructure and rephrase them in the best English possible.

For example, if we ask ChatGPT to provide a basic introduction to Groovy with examples of accessing getters and setters as properties and showing collection shortcuts, it can give us a well-written and easy-to-understand answer.

To make the most of this tool, I'm calling for a session where we can generate a huge amount of high quality documentation in a short time. Before we get started, let's collect ideas for what should be documented. Please add your suggestions as comments to this post.

Remember, the content and ideas are still our responsibility as humans. But ChatGPT can help us with the "how" – at least as long as it remains available for free or at an affordable price. Let's take advantage of this amazing resource and improve our documentation for Freeplane features.

**Comment by dpolivaev** [comment]
For example, if I ask it "give me some basic introduction to Groovy with examples accessing getters and setters as properties and showing collection shortcuts" it answers

--------- 

One of the key features of Groovy is its ability to treat getters and setters as properties, which means you can access them using the dot notation like you would with any other object property. For example, consider the following Java class:

~~~~java
public class Person {
  private String name;
  private int age;

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  public int getAge() {
    return age;
  }

  public void setAge(int age) {
    this.age = age;
  }
}

~~~~~

In Groovy, you could access the name and age properties of a Person object like this:

~~~~~groovy
def person = new Person()
person.name = "John"
person.age = 30

println "Name: ${person.name}"
println "Age: ${person.age}"

~~~~~

Groovy also offers a number of shortcuts for working with collections, such as lists and maps. For example, you can create a list using the [...] notation, like this:

~~~~~groovy
def list = ['apple', 'banana', 'orange']
~~~~~

You can access elements of the list using the index notation, like this:

~~~~~groovy
def firstElement = list[0]
~~~~~

You can also use the each method to iterate over the elements of the list, like this:

~~~~~groovy
list.each { element ->
  println element
}
~~~~~

**Comment by euu2021** [comment]
I did some tests and I see that ChatGPT can be useful to write the documentation, indeed. 

For the part about scripts in the documentation, which is the part that I'm interested right now, I still need to understand it better, before I can use ChatGPT to generate content. I already completed an overview about Java basics, and now I'm learning about how the API/Proxy of Freeplane works.

**Comment by macmarrum** [comment]
As far as I understand Freeplane documentation is covered by the same license as Freeplane - GPLv2

I wonder whether the license is compatible with content generated by ChatGPT.
___

ChatGPT

Q: What license applies to content generated as answers?

A: The content that I generate as an AI assistant is provided for general information purposes only. I am not a lawyer and cannot provide legal advice. It is your responsibility to determine whether the content I provide is suitable for your needs and to obtain any necessary permissions or licenses that may be required to use or distribute the content. OpenAI, the company that developed me, holds the copyright to the software that I am built on and to the content that I generate. However, as a general rule, OpenAI allows the use of my content for non-commercial purposes as long as proper attribution is given. Please see OpenAI's terms of use for more information.

https://openai.com/terms/
> 3. Content
(a) Your Content. You may provide input to the Services (“Input”), and receive output generated and returned by the Services based on the Input (“Output”). Input and Output are collectively “Content.” As between the parties and to the extent permitted by applicable law, you own all Input, and subject to your compliance with these Terms, OpenAI hereby assigns to you all its right, title and interest in and to Output. OpenAI may use Content as necessary to provide and maintain the Services, comply with applicable law, and enforce our policies. You are responsible for Content, including for ensuring that it does not violate any applicable law or these Terms.

**Reply by dpolivaev** [reply]
I am sure that this limitation does not apply to our documentation pages because they are not distributed with the software.
Further, even if we want to distribute mind maps containing texts produced by the AI, they are not a part of the program code and therefore we would have no collisions with GPL. 

According to the quote, we are assigned all rights to the output. Therefore I do not see any legal complications coming.

**Reply by macmarrum** [reply]
> I am sure that this limitation does not apply to our documentation pages because they are not distributed with the software.

I am not sure what you mean by that. I can think of two meanings:
1. Freeplane documentation is not covered by GPLv2 because it is not distributed
2. Freeplane documentation is not covered by GPLv2 because it is not part of the software

**As to 1**
When something is made available on a web server, like GitHub, and someone downloads it to their device, like you just did with this very post, distribution occurs.
There is no doubt in my mind that Freeplane documentation is distributed.

**As to 2**
Freeplane documentation is made available alongside Freeplane, on GitHub. Freeplane is GPLv2-licensed. There is no other license specified for the documentation. It follows that Feeplane documentation is also GPLv2-licensed.

> According to the quote, we are assigned all rights to the output. 

In **OpenAI's terms of use** I can see: "_OpenAI hereby assigns to you all its right [...] to Output_", i.e. only the right that it already has. It is possible that Output includes content to which OpenAI has no right or the right of the original author is infringed.
The terms mention further that OpenAI "_disclaim all warranties including but not limited to warranties of [...] non-infringement_".

**Reply by dpolivaev** [reply]
Please read an excerpt of the conversation I just had with ChatGPT. Personally, I am convinced by the explanations and reasons it delivered.

------

Me: Who is the author of texts generated with your assistance?

AI: As an artificial intelligence, I am not the author of texts generated with my assistance. The author of a text is the person or entity who creates the text and is responsible for its content.

In the case of texts generated with my assistance, the author would generally be the person or entity that provided the input and directed the content of the text. This could be an individual user or a team of writers working together.

It's important to note that, even though I may have helped to generate the text, the ultimate responsibility for its content and accuracy lies with the person or entity that directed its creation.

If the text is being created as part of a collaboration, it's important to clarify in advance who will be credited as the author and how the text will be used and distributed. This can be done through a written agreement or contract.

Me: Can texts generated with your assistance be used in the software documentation documents if they are distributed with software is licensed under GPL2+?

AI: Yes, you can use texts generated with my assistance in software documentation documents if they are distributed with software that is licensed under GPL2+. GPL2+ is a widely used open source software license that allows users to freely modify, distribute, and use the software, as long as they provide credit to the original creators and distribute the modified version under the same license.

As long as you follow the terms of the GPL2+ license, you can use texts generated with my assistance in your software documentation. This includes properly attributing any text that you use, and making sure that the documentation is also distributed under the GPL2+ license or a compatible license.

It is always a good idea to carefully read and understand the terms of any license that you are using, and to seek legal advice if you have any questions or concerns about how to comply with the terms of the license.

**Reply by sschwarzer** [reply]
@dpolivaev Maybe we're already on the same page, but just in case: Although I agree that the ChatGPT output sounds reasonable, I'd like to point out that, as "just" a very sophisticated chatbot, ChatGPT can't legally represent OpenAI (or anyone else). :-)

That said, I think as long as the ChatGPT output for the documentation is just the original documentation in a more polished style, I wouldn't expect problems. On the other hand, if the AI outputs additional content, it's unclear where it comes from and whether that content may be distributed.
